docs/content/preview/contribute/core-database/build-and-test.md

```diff
@@ -0,0 +1,251 @@
+
+### Third-party
+
+By default, third-party libraries are pre-built into archives, and those archives are downloaded to be used during build.
+Incremental builds currently do not detect when the third-party archive URL has been updated, so when that happens, you should run a `--clean` build to use the new third-party archive.
+You can also build the third-party archives locally using `--no-download-thirdparty` or `--ndltp`.
+
+## Test
+
+The following examples use the `release` build type, but any build type can be specified, provided that build type is not disabled for that test.
+For example, several tests are disabled for the `tsan` build type.
+
+### C++ tests
+
+#### Run all tests
+
+To run all the C++ tests, run the following command:
+
+```sh
+./yb_build.sh release --ctest
+```
+
+#### Run specific tests
+
+To run a specific test, for example the `util_monotime-test` test, run the following command:
+
+```sh
+./yb_build.sh release --cxx-test util_monotime-test
+```
+
+To run a specific sub-test, for example the `TestMonoTime.TestCondition` sub-test in `util_monotime-test`, run the following command:
+
+```sh
+./yb_build.sh release --cxx-test util_monotime-test --gtest_filter TestMonoTime.TestCondition
+```
+
+`--gtest_filter` supports globbing.
+For example, `TestMonoTime.TestTime*` runs both `TestMonoTime.TestTimeVal` and `TestMonoTime.TestTimeSpec`.
+Make sure to escape or quote the `*` if your shell interprets it as a glob character.
+
+### Java tests
+
+#### Run all tests
+
+To run all the Java tests, use the following command:
+
+```sh
+./yb_build.sh release --java-tests
+```
+
+#### Run specific tests
+
+To run a specific test:
+
+```sh
+./yb_build.sh release --java-test org.yb.client.TestYBClient
+```
+
+To run a specific Java sub-test within a test file, use the # syntax.
+For example:
+
+```sh
+./yb_build.sh release --java-test 'org.yb.client.TestYBClient#testClientCreateDestroy'
+```
+
+`--scb` and `--sj` are recommended in case the C++ and Java code was already built beforehand.
+
+### Run tests multiple times
+
+`yb_build.sh` has a built-in capability of running tests multiple times with parallelism.
+
+- `--num_repetitions`/`-n`: number of times to run the test(s)
+- `--test-parallelism`/`--tp`: number of instances of the test to run in parallel
+
+Note that a high parallelism could result in failures if system resources are overused.
+
+### Test frameworks
+
+#### C++ (external) mini cluster
+
+C++ tests are located in `src/yb` or `ent/src/yb` and end in `test.cc`.
+
+Many C++ tests use external mini cluster or mini cluster to create clusters.
+A key difference is that external mini clusters create separate master, tserver, and postgres processes (as in real life), while mini clusters create masters and tservers in memory in the same process and optionally spawn a separate postgres process.
```

Comment:
Not sure what "in memory" means here. These in-process masters/tservers use disk similarly to a real server process.